File: backend/internal/awsx/cloudwatch.go
package awsx

// CloudWatch metric fetch + summarisation, ported from cloudwatch.ts.

import (
	"context"
	"fmt"
	"math"
	"sort"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/cloudwatch"
	cwtypes "github.com/aws/aws-sdk-go-v2/service/cloudwatch/types"

	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/config"
	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/store"
	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/types"
)

type rawPoint struct {
	t int64
	v float64
}

type RawSeries struct {
	Key    string
	Label  string
	Unit   string
	Stat   string // "Average" | "Sum"
	Points []rawPoint
	// ThresholdKey lets a per-TG series borrow the shared threshold.
	ThresholdKey string
	// Metric is the CloudWatch metric this came from — what someone checking
	// the number in the console has to search for.
	Metric string
}

func toSeries(key, label, unit, stat, metric string, r *cwtypes.MetricDataResult, thresholdKey string) RawSeries {
	points := []rawPoint{}
	if r != nil {
		for i, t := range r.Timestamps {
			v := 0.0
			if i < len(r.Values) {
				v = r.Values[i]
			}
			points = append(points, rawPoint{t: t.UnixMilli(), v: v})
		}
		sort.Slice(points, func(i, j int) bool { return points[i].t < points[j].t })
	}
	return RawSeries{Key: key, Label: label, Unit: unit, Stat: stat, Points: points, ThresholdKey: thresholdKey, Metric: metric}
}

// aggBuckets buckets are aggregated into the headline number to smooth a
// single noisy bucket; every summary carries a basis saying what it counted.
const aggBuckets = 3

func round3f(n float64) float64 { return math.Round(n*1000) / 1000 }

func Summarize(s RawSeries, win types.ResolvedWindow) types.MetricSummary {
	// Drop the newest (possibly incomplete) bucket, then compare the last 3
	// complete buckets against the 3 before them.
	pts := s.Points
	if len(pts) > 0 {
		pts = pts[:len(pts)-1]
	}
	slice := func(from, to int) []rawPoint {
		if from < 0 {
			from = 0
		}
		if to < from {
			to = from
		}
		return pts[from:to]
	}
	currentWin := slice(len(pts)-aggBuckets, len(pts))
	prevWin := slice(len(pts)-2*aggBuckets, len(pts)-aggBuckets)
	// A Sum metric is normalised to a per-minute rate rather than left as a
	// bucket total — thresholds are absolute, so a raw total would make the
	// alert depend on the chosen interval.
	agg := func(w []rawPoint) float64 {
		if len(w) == 0 {
			return 0
		}
		sum := 0.0
		for _, p := range w {
			sum += p.v
		}
		if s.Stat == "Average" {
			return sum / float64(len(w))
		}
		return sum / (float64(len(w)) * float64(win.IntervalMin))
	}
	current := round3f(agg(currentWin))
	previous := round3f(agg(prevWin))
	delta := round3f(current - previous)
	var percentChange *float64
	switch {
	case previous > 0:
		percentChange = types.Ptr(round3f((current - previous) / previous * 100))
	case current > 0:
		percentChange = nil
	default:
		percentChange = types.Ptr(0.0)
	}
	points := make([]types.MetricPoint, 0, len(pts))
	for _, p := range pts {
		points = append(points, types.MetricPoint{
			T: time.UnixMilli(p.t).UTC().Format(time.RFC3339Nano),
			V: round3f(p.v),
		})
	}
	thresholdKey := s.ThresholdKey
	if thresholdKey == "" {
		thresholdKey = s.Key
	}
	basis := fmt.Sprintf("%s %s · ", s.Metric, s.Stat)
	if s.Stat == "Sum" {
		basis += fmt.Sprintf("최근 %d버킷(%d분) 합계를 분당으로 환산 · 직전 동일 구간과 비교", aggBuckets, aggBuckets*win.IntervalMin)
	} else {
		basis += fmt.Sprintf("최근 %d버킷(%d분) 평균 · 직전 동일 구간과 비교", aggBuckets, aggBuckets*win.IntervalMin)
	}
	return types.MetricSummary{
		Key:           s.Key,
		Label:         s.Label,
		Unit:          s.Unit,
		Current:       current,
		Previous:      previous,
		Delta:         delta,
		PercentChange: percentChange,
		Status:        config.StatusFor(thresholdKey, current, percentChange),
		Points:        points,
		Basis:         basis,
	}
}

type CoreMetricsResult struct {
	Summaries []types.MetricSummary
	Errors    []string
}

func q(id, namespace, metricName string, dims []cwtypes.Dimension, stat string, periodSec int32) cwtypes.MetricDataQuery {
	return cwtypes.MetricDataQuery{
		Id: aws.String(id),
		MetricStat: &cwtypes.MetricStat{
			Metric: &cwtypes.Metric{
				Namespace:  aws.String(namespace),
				MetricName: aws.String(metricName),
				Dimensions: dims,
			},
			Period: aws.Int32(periodSec),
			Stat:   aws.String(stat),
		},
		ReturnData: aws.Bool(true),
	}
}

func dim(name, value string) cwtypes.Dimension {
	return cwtypes.Dimension{Name: aws.String(name), Value: aws.String(value)}
}

func byID(results []cwtypes.MetricDataResult) map[string]*cwtypes.MetricDataResult {
	out := map[string]*cwtypes.MetricDataResult{}
	for i := range results {
		out[aws.ToString(results[i].Id)] = &results[i]
	}
	return out
}

func (a *AWS) FetchCoreMetrics(ctx context.Context, win types.ResolvedWindow) (CoreMetricsResult, error) {
	end := time.UnixMilli(win.EndMs)
	start := time.UnixMilli(win.StartMs)
	periodSec := int32(win.IntervalMin * 60)

	results := []RawSeries{}
	errors := []string{}

	// --- ALB + RDS Proxy (workload region) ---
	if err := func() error {
		alb, err := a.DiscoverAlb(ctx)
		if err != nil {
			return err
		}
		albDim := []cwtypes.Dimension{dim("LoadBalancer", alb.LoadBalancer)}
		rdsProxy := a.Settings.RdsProxyName()
		queries := []cwtypes.MetricDataQuery{
			q("trt", "AWS/ApplicationELB", "TargetResponseTime", albDim, "Average", periodSec),
			q("c4xx", "AWS/ApplicationELB", "HTTPCode_Target_4XX_Count", albDim, "Sum", periodSec),
			q("c5xx", "AWS/ApplicationELB", "HTTPCode_Target_5XX_Count", albDim, "Sum", periodSec),
			q("c2xx", "AWS/ApplicationELB", "HTTPCode_Target_2XX_Count", albDim, "Sum", periodSec),
			q("c3xx", "AWS/ApplicationELB", "HTTPCode_Target_3XX_Count", albDim, "Sum", periodSec),
			q("reqs", "AWS/ApplicationELB", "RequestCount", albDim, "Sum", periodSec),
			q("rdscc", "AWS/RDS", "ClientConnections", []cwtypes.Dimension{dim("ProxyName", rdsProxy)}, "Average", periodSec),
			q("rdsdc", "AWS/RDS", "DatabaseConnections", []cwtypes.Dimension{dim("ProxyName", rdsProxy)}, "Average", periodSec),
		}
		client, err := a.cloudWatch(ctx, a.Settings.Region())
		if err != nil {
			return err
		}
		res, err := client.GetMetricData(ctx, &cloudwatch.GetMetricDataInput{
			StartTime: aws.Time(start), EndTime: aws.Time(end), MetricDataQueries: queries,
		})
		if err != nil {
			return err
		}
		m := byID(res.MetricDataResults)
		results = append(results,
			toSeries("targetResponseTime", "TargetResponseTime", "s", "Average", "AWS/ApplicationELB TargetResponseTime", m["trt"], ""),
			toSeries("http4xx", "Target 4XX", "req/min", "Sum", "AWS/ApplicationELB HTTPCode_Target_4XX_Count", m["c4xx"], ""),
			toSeries("http5xx", "Target 5XX", "req/min", "Sum", "AWS/ApplicationELB HTTPCode_Target_5XX_Count", m["c5xx"], ""),
			toSeries("http2xx", "Target 2XX", "req/min", "Sum", "AWS/ApplicationELB HTTPCode_Target_2XX_Count", m["c2xx"], ""),
			toSeries("http3xx", "Target 3XX", "req/min", "Sum", "AWS/ApplicationELB HTTPCode_Target_3XX_Count", m["c3xx"], ""),
			toSeries("requestCount", "RequestCount", "req/min", "Sum", "AWS/ApplicationELB RequestCount", m["reqs"], ""),
			toSeries("rdsClientConnections", "RDS Proxy Client Conn", "conn", "Average", fmt.Sprintf("AWS/RDS ClientConnections (ProxyName=%s)", rdsProxy), m["rdscc"], ""),
			toSeries("rdsDatabaseConnections", "RDS Proxy DB Conn", "conn", "Average", fmt.Sprintf("AWS/RDS DatabaseConnections (ProxyName=%s)", rdsProxy), m["rdsdc"], ""),
		)
		return nil
	}(); err != nil {
		errors = append(errors, "ALB/RDS metrics: "+ErrMsg(err))
	}

	// --- WAF Blocked/Allowed (us-east-1 for CLOUDFRONT scope) ---
	if err := func() error {
		aclName := a.Settings.WafWebAclName()
		dims := []cwtypes.Dimension{dim("WebACL", aclName), dim("Rule", "ALL")}
		if a.Settings.WafScope() != "CLOUDFRONT" {
			dims = append(dims, dim("Region", a.Settings.Region()))
		}
		client, err := a.cloudWatch(ctx, a.Settings.WafRegion())
		if err != nil {
			return err
		}
		res, err := client.GetMetricData(ctx, &cloudwatch.GetMetricDataInput{
			StartTime: aws.Time(start), EndTime: aws.Time(end),
			MetricDataQueries: []cwtypes.MetricDataQuery{
				q("wafb", "AWS/WAFV2", "BlockedRequests", dims, "Sum", periodSec),
				q("wafa", "AWS/WAFV2", "AllowedRequests", dims, "Sum", periodSec),
			},
		})
		if err != nil {
			return err
		}
		m := byID(res.MetricDataResults)
		results = append(results,
			toSeries("wafBlocked", "WAF BlockedRequests", "req/min", "Sum", fmt.Sprintf("AWS/WAFV2 BlockedRequests (WebACL=%s, Rule=ALL)", aclName), m["wafb"], ""),
			toSeries("wafAllowed", "WAF AllowedRequests", "req/min", "Sum", fmt.Sprintf("AWS/WAFV2 AllowedRequests (WebACL=%s, Rule=ALL)", aclName), m["wafa"], ""),
		)
		return nil
	}(); err != nil {
		errors = append(errors, "WAF metrics: "+ErrMsg(err))
	}

	if len(results) == 0 {
		msg := "no metric data"
		if len(errors) > 0 {
			msg = ""
			for i, e := range errors {
				if i > 0 {
					msg += " / "
				}
				msg += e
			}
		}
		return CoreMetricsResult{}, fmt.Errorf("%s", msg)
	}

	for _, s := range results {
		// metric cache failure must not break the panel
		samples := make([]store.Sample, 0, len(s.Points))
		for _, p := range s.Points {
			samples = append(samples, store.Sample{T: p.t, V: p.v})
		}
		_ = a.Store.SaveMetricSamples(s.Key, samples)
	}

	summaries := make([]types.MetricSummary, 0, len(results))
	for _, r := range results {
		summaries = append(summaries, Summarize(r, win))
	}
	return CoreMetricsResult{Summaries: summaries, Errors: errors}, nil
}

// FetchTargetGroupMetrics: per-Target-Group ALB metrics (spec item 3).
func (a *AWS) FetchTargetGroupMetrics(ctx context.Context, win types.ResolvedWindow) ([]types.TargetGroupMetrics, error) {
	end := time.UnixMilli(win.EndMs)
	start := time.UnixMilli(win.StartMs)
	periodSec := int32(win.IntervalMin * 60)
	alb, err := a.DiscoverAlb(ctx)
	if err != nil {
		return nil, err
	}
	if len(alb.TargetGroups) == 0 {
		return []types.TargetGroupMetrics{}, nil
	}

	queries := []cwtypes.MetricDataQuery{}
	for i, tg := range alb.TargetGroups {
		dims := []cwtypes.Dimension{dim("LoadBalancer", alb.LoadBalancer), dim("TargetGroup", tg.TgDim)}
		queries = append(queries,
			q(fmt.Sprintf("tg%dtrt", i), "AWS/ApplicationELB", "TargetResponseTime", dims, "Average", periodSec),
			q(fmt.Sprintf("tg%dc4", i), "AWS/ApplicationELB", "HTTPCode_Target_4XX_Count", dims, "Sum", periodSec),
			q(fmt.Sprintf("tg%dc5", i), "AWS/ApplicationELB", "HTTPCode_Target_5XX_Count", dims, "Sum", periodSec),
		)
	}
	client, err := a.cloudWatch(ctx, a.Settings.Region())
	if err != nil {
		return nil, err
	}
	res, err := client.GetMetricData(ctx, &cloudwatch.GetMetricDataInput{
		StartTime: aws.Time(start), EndTime: aws.Time(end), MetricDataQueries: queries,
	})
	if err != nil {
		return nil, err
	}
	m := byID(res.MetricDataResults)

	out := make([]types.TargetGroupMetrics, 0, len(alb.TargetGroups))
	for i, tg := range alb.TargetGroups {
		trt := Summarize(toSeries(
			fmt.Sprintf("tg-%s-trt", tg.Name), "TargetResponseTime", "s", "Average",
			fmt.Sprintf("AWS/ApplicationELB TargetResponseTime (TargetGroup=%s)", tg.Name),
			m[fmt.Sprintf("tg%dtrt", i)], "targetResponseTime"), win)
		c4 := Summarize(toSeries(
			fmt.Sprintf("tg-%s-4xx", tg.Name), "4XX", "req/min", "Sum",
			fmt.Sprintf("AWS/ApplicationELB HTTPCode_Target_4XX_Count (TargetGroup=%s)", tg.Name),
			m[fmt.Sprintf("tg%dc4", i)], "http4xx"), win)
		c5 := Summarize(toSeries(
			fmt.Sprintf("tg-%s-5xx", tg.Name), "5XX", "req/min", "Sum",
			fmt.Sprintf("AWS/ApplicationELB HTTPCode_Target_5XX_Count (TargetGroup=%s)", tg.Name),
			m[fmt.Sprintf("tg%dc5", i)], "http5xx"), win)
		out = append(out, types.TargetGroupMetrics{
			Name: tg.Name, PathPattern: tg.PathPattern, ResponseTime: trt, C4xx: c4, C5xx: c5,
		})
	}
	return out, nil
}

File: backend/awsx/cloudwatch.ts
// CloudWatch metric fetch + summarisation.

import {
  GetMetricDataCommand,
  type Dimension,
  type MetricDataQuery,
  type MetricDataResult,
} from "@aws-sdk/client-cloudwatch";

import { statusFor } from "../config/thresholds.ts";
import type { Sample } from "../store/store.ts";
import type {
  MetricPoint,
  MetricSummary,
  ResolvedWindow,
  TargetGroupMetrics,
} from "../../src/lib/types.ts";
import { AWS, errMsg } from "./clients.ts";
import { discoverAlb } from "./alb.ts";

interface RawPoint {
  t: number;
  v: number;
}

export interface RawSeries {
  key: string;
  label: string;
  unit: string;
  /** "Average" | "Sum" */
  stat: string;
  points: RawPoint[];
  /** Lets a per-TG series borrow the shared threshold. */
  thresholdKey: string;
  /**
   * The CloudWatch metric this came from — what someone checking the number in
   * the console has to search for.
   */
  metric: string;
}

function toSeries(
  key: string,
  label: string,
  unit: string,
  stat: string,
  metric: string,
  r: MetricDataResult | undefined,
  thresholdKey: string,
): RawSeries {
  const points: RawPoint[] = [];
  if (r?.Timestamps) {
    r.Timestamps.forEach((t, i) => {
      points.push({ t: new Date(t).getTime(), v: r.Values?.[i] ?? 0 });
    });
    points.sort((a, b) => a.t - b.t);
  }
  return { key, label, unit, stat, points, thresholdKey, metric };
}

/**
 * How many buckets are aggregated into the headline number, to smooth a single
 * noisy bucket; every summary carries a basis saying what it counted.
 */
const AGG_BUCKETS = 3;

function round3(n: number): number {
  return Math.round(n * 1000) / 1000;
}

export function summarize(s: RawSeries, win: ResolvedWindow): MetricSummary {
  // Drop the newest (possibly incomplete) bucket, then compare the last 3
  // complete buckets against the 3 before them.
  const pts = s.points.length > 0 ? s.points.slice(0, -1) : s.points;
  const slice = (from: number, to: number) => pts.slice(Math.max(from, 0), Math.max(to, 0));
  const currentWin = slice(pts.length - AGG_BUCKETS, pts.length);
  const prevWin = slice(pts.length - 2 * AGG_BUCKETS, pts.length - AGG_BUCKETS);

  // A Sum metric is normalised to a per-minute rate rather than left as a bucket
  // total — thresholds are absolute, so a raw total would make the alert depend
  // on the chosen interval.
  const agg = (w: RawPoint[]): number => {
    if (w.length === 0) return 0;
    const sum = w.reduce((acc, p) => acc + p.v, 0);
    if (s.stat === "Average") return sum / w.length;
    return sum / (w.length * win.intervalMin);
  };

  const current = round3(agg(currentWin));
  const previous = round3(agg(prevWin));
  const delta = round3(current - previous);
  let percentChange: number | null;
  if (previous > 0) percentChange = round3(((current - previous) / previous) * 100);
  else if (current > 0) percentChange = null;
  else percentChange = 0;

  const points: MetricPoint[] = pts.map((p) => ({
    t: new Date(p.t).toISOString(),
    v: round3(p.v),
  }));

  const thresholdKey = s.thresholdKey || s.key;
  const span = AGG_BUCKETS * win.intervalMin;
  const basis =
    `${s.metric} ${s.stat} · ` +
    (s.stat === "Sum"
      ? `최근 ${AGG_BUCKETS}버킷(${span}분) 합계를 분당으로 환산 · 직전 동일 구간과 비교`
      : `최근 ${AGG_BUCKETS}버킷(${span}분) 평균 · 직전 동일 구간과 비교`);

  return {
    key: s.key,
    label: s.label,
    unit: s.unit,
    current,
    previous,
    delta,
    percentChange,
    status: statusFor(thresholdKey, current, percentChange),
    points,
    basis,
  };
}

export interface CoreMetricsResult {
  summaries: MetricSummary[];
  errors: string[];
}

function q(
  id: string,
  namespace: string,
  metricName: string,
  dims: Dimension[],
  stat: string,
  periodSec: number,
): MetricDataQuery {
  return {
    Id: id,
    MetricStat: {
      Metric: { Namespace: namespace, MetricName: metricName, Dimensions: dims },
      Period: periodSec,
      Stat: stat,
    },
    ReturnData: true,
  };
}

function dim(name: string, value: string): Dimension {
  return { Name: name, Value: value };
}

function byId(results: MetricDataResult[] | undefined): Map<string, MetricDataResult> {
  const out = new Map<string, MetricDataResult>();
  for (const r of results ?? []) {
    if (r.Id) out.set(r.Id, r);
  }
  return out;
}

export async function fetchCoreMetrics(
  a: AWS,
  win: ResolvedWindow,
): Promise<CoreMetricsResult> {
  const end = new Date(win.endMs);
  const start = new Date(win.startMs);
  const periodSec = win.intervalMin * 60;

  const results: RawSeries[] = [];
  const errors: string[] = [];

  // --- ALB + RDS Proxy (workload region) ---
  try {
    const alb = await discoverAlb(a);
    const albDim = [dim("LoadBalancer", alb.loadBalancer)];
    const rdsProxy = a.settings.rdsProxyName();
    const queries = [
      q("trt", "AWS/ApplicationELB", "TargetResponseTime", albDim, "Average", periodSec),
      q("c4xx", "AWS/ApplicationELB", "HTTPCode_Target_4XX_Count", albDim, "Sum", periodSec),
      q("c5xx", "AWS/ApplicationELB", "HTTPCode_Target_5XX_Count", albDim, "Sum", periodSec),
      q("c2xx", "AWS/ApplicationELB", "HTTPCode_Target_2XX_Count", albDim, "Sum", periodSec),
      q("c3xx", "AWS/ApplicationELB", "HTTPCode_Target_3XX_Count", albDim, "Sum", periodSec),
      q("reqs", "AWS/ApplicationELB", "RequestCount", albDim, "Sum", periodSec),
      q("rdscc", "AWS/RDS", "ClientConnections", [dim("ProxyName", rdsProxy)], "Average", periodSec),
      q("rdsdc", "AWS/RDS", "DatabaseConnections", [dim("ProxyName", rdsProxy)], "Average", periodSec),
    ];
    const res = await a.cloudWatch(a.settings.region()).send(
      new GetMetricDataCommand({
        StartTime: start,
        EndTime: end,
        MetricDataQueries: queries,
      }),
    );
    const m = byId(res.MetricDataResults);
    results.push(
      toSeries("targetResponseTime", "TargetResponseTime", "s", "Average", "AWS/ApplicationELB TargetResponseTime", m.get("trt"), ""),
      toSeries("http4xx", "Target 4XX", "req/min", "Sum", "AWS/ApplicationELB HTTPCode_Target_4XX_Count", m.get("c4xx"), ""),
      toSeries("http5xx", "Target 5XX", "req/min", "Sum", "AWS/ApplicationELB HTTPCode_Target_5XX_Count", m.get("c5xx"), ""),
      toSeries("http2xx", "Target 2XX", "req/min", "Sum", "AWS/ApplicationELB HTTPCode_Target_2XX_Count", m.get("c2xx"), ""),
      toSeries("http3xx", "Target 3XX", "req/min", "Sum", "AWS/ApplicationELB HTTPCode_Target_3XX_Count", m.get("c3xx"), ""),
      toSeries("requestCount", "RequestCount", "req/min", "Sum", "AWS/ApplicationELB RequestCount", m.get("reqs"), ""),
      toSeries("rdsClientConnections", "RDS Proxy Client Conn", "conn", "Average", `AWS/RDS ClientConnections (ProxyName=${rdsProxy})`, m.get("rdscc"), ""),
      toSeries("rdsDatabaseConnections", "RDS Proxy DB Conn", "conn", "Average", `AWS/RDS DatabaseConnections (ProxyName=${rdsProxy})`, m.get("rdsdc"), ""),
    );
  } catch (e) {
    errors.push("ALB/RDS metrics: " + errMsg(e));
  }

  // --- WAF Blocked/Allowed (us-east-1 for CLOUDFRONT scope) ---
  try {
    const aclName = a.settings.wafWebAclName();
    const dims = [dim("WebACL", aclName), dim("Rule", "ALL")];
    if (a.settings.wafScope() !== "CLOUDFRONT") dims.push(dim("Region", a.settings.region()));
    const res = await a.cloudWatch(a.settings.wafRegion()).send(
      new GetMetricDataCommand({
        StartTime: start,
        EndTime: end,
        MetricDataQueries: [
          q("wafb", "AWS/WAFV2", "BlockedRequests", dims, "Sum", periodSec),
          q("wafa", "AWS/WAFV2", "AllowedRequests", dims, "Sum", periodSec),
        ],
      }),
    );
    const m = byId(res.MetricDataResults);
    results.push(
      toSeries("wafBlocked", "WAF BlockedRequests", "req/min", "Sum", `AWS/WAFV2 BlockedRequests (WebACL=${aclName}, Rule=ALL)`, m.get("wafb"), ""),
      toSeries("wafAllowed", "WAF AllowedRequests", "req/min", "Sum", `AWS/WAFV2 AllowedRequests (WebACL=${aclName}, Rule=ALL)`, m.get("wafa"), ""),
    );
  } catch (e) {
    errors.push("WAF metrics: " + errMsg(e));
  }

  if (results.length === 0) {
    throw new Error(errors.length > 0 ? errors.join(" / ") : "no metric data");
  }

  for (const s of results) {
    // A metric-cache failure must not break the panel.
    try {
      const samples: Sample[] = s.points.map((p) => ({ t: p.t, v: p.v }));
      a.store?.saveMetricSamples(s.key, samples);
    } catch {
      // ignored
    }
  }

  return { summaries: results.map((r) => summarize(r, win)), errors };
}

/** Per-Target-Group ALB metrics (spec item 3). */
export async function fetchTargetGroupMetrics(
  a: AWS,
  win: ResolvedWindow,
): Promise<TargetGroupMetrics[]> {
  const end = new Date(win.endMs);
  const start = new Date(win.startMs);
  const periodSec = win.intervalMin * 60;
  const alb = await discoverAlb(a);
  if (alb.targetGroups.length === 0) return [];

  const queries: MetricDataQuery[] = [];
  alb.targetGroups.forEach((tg, i) => {
    const dims = [dim("LoadBalancer", alb.loadBalancer), dim("TargetGroup", tg.tgDim)];
    queries.push(
      q(`tg${i}trt`, "AWS/ApplicationELB", "TargetResponseTime", dims, "Average", periodSec),
      q(`tg${i}c4`, "AWS/ApplicationELB", "HTTPCode_Target_4XX_Count", dims, "Sum", periodSec),
      q(`tg${i}c5`, "AWS/ApplicationELB", "HTTPCode_Target_5XX_Count", dims, "Sum", periodSec),
    );
  });

  const res = await a.cloudWatch(a.settings.region()).send(
    new GetMetricDataCommand({ StartTime: start, EndTime: end, MetricDataQueries: queries }),
  );
  const m = byId(res.MetricDataResults);

  return alb.targetGroups.map((tg, i) => ({
    name: tg.name,
    pathPattern: tg.pathPattern,
    responseTime: summarize(
      toSeries(
        `tg-${tg.name}-trt`, "TargetResponseTime", "s", "Average",
        `AWS/ApplicationELB TargetResponseTime (TargetGroup=${tg.name})`,
        m.get(`tg${i}trt`), "targetResponseTime",
      ),
      win,
    ),
    c4xx: summarize(
      toSeries(
        `tg-${tg.name}-4xx`, "4XX", "req/min", "Sum",
        `AWS/ApplicationELB HTTPCode_Target_4XX_Count (TargetGroup=${tg.name})`,
        m.get(`tg${i}c4`), "http4xx",
      ),
      win,
    ),
    c5xx: summarize(
      toSeries(
        `tg-${tg.name}-5xx`, "5XX", "req/min", "Sum",
        `AWS/ApplicationELB HTTPCode_Target_5XX_Count (TargetGroup=${tg.name})`,
        m.get(`tg${i}c5`), "http5xx",
      ),
      win,
    ),
  }));
}
